icu4c/source/common/unistr.cpp

```diff
@@ -1465,17 +1470,28 @@ UnicodeString::doReplace(int32_t start,
   }
   newLength += srcLength;
 
+  // Check for insertion into ourself
+  const UChar *oldArray = getArrayStart();
+  if (isBufferWritable() &&
+      oldArray < srcChars + srcLength &&
```

Comment:
The pointer comparisons are technically undefined behavior if oldArray and srcChars are not part of the same array/allocation. Although there is nothing much practical to be done about it, and we assume it works everywhere else. (I'm not suggesting any change.)